daliuge-engine/dlg/io.py

```diff
@@ -240,18 +321,39 @@ class MemoryIO(DataIO):
     A DataIO class that reads/write from/into the BytesIO object given at
     construction time
     """
-    _desc: io.BytesIO  # TODO: This might actually be a problem
+    # TODO: This might actually be a problem
+    _desc: io.BufferedReader
+    _buf: io.BytesIO
 
     def __init__(self, buf: io.BytesIO, **kwargs):
         super().__init__()
         self._buf = buf
 
+    class BytesIOReader(io.BufferedReader):
+        """
+        An BinaryIO Reader that wraps a BytesIO object for concurrent
+        reading and writing. Closing this reader will not close other
+        readers observing the same BytesIO object.
+        """
+        _closed = False
+        def __init__(self, raw: io.BytesIO, buffer_size: int = 2048):
+            assert isinstance(raw, io.BytesIO)
+            # NOTE: BytesIO extends BufferedIOBase instead of RawIOBase. Read
+            # and peek operations may return more bytes than requested.
+            super().__init__(raw, buffer_size)  # type: ignore
+        def close(self) -> None:
+            self._closed = True
+        @property
+        def closed(self) -> bool:
+            return self.closed
+
     def _open(self, **kwargs):
         if self._mode == OpenMode.OPEN_WRITE:
             return self._buf
         elif self._mode == OpenMode.OPEN_READ:
-            # TODO: potentially wasteful copy
-            return io.BytesIO(self._buf.getbuffer())
+            br = MemoryIO.BytesIOReader(self._buf)
```

Comment:
One of the bugs I've noticed around this that numerous read functions call close on the io object after reading. When updating such that multiple readers all point to the same _buf object, the reader closing additionally closes the bytesio object.

DataDrop should still be correctly handling the closing of the underlying BytesIO object.